source/core/shape.cc

```diff
@@ -363,6 +363,195 @@ Sphere<dim>::set_position(const Point<dim> &position)
+          distance_gradient = superquadric_gradient(current_point);
+          if (distance_gradient.norm() < epsilon)
+            // Gradient can be null if the evaluation point is exactly on the
+            // centroid of the shape. In this case it is also the closest point.
+            break;
+          dx = -relaxation * (current_distance * distance_gradient) /
+               distance_gradient.norm_square();
+
+          current_point    = current_point + dx;
+          current_distance = superquadric(current_point);
+
+          iteration++;
+        }
+
+      closest_point = this->reverse_align_and_center(current_point);
+    }
+  else
+    closest_point = iterator->second;
+}
+
+template <int dim>
+double
+Superquadric<dim>::value(const Point<dim> &evaluation_point,
+                         const unsigned int /*component*/) const
+{
+  using numbers::PI;
+
+  auto point_in_string = this->point_to_string(evaluation_point);
+  auto iterator        = this->value_cache.find(point_in_string);
+  if (iterator == this->value_cache.end())
+    {
+      // The closest point has to be found first, because it is used in value
+      // calculation.
+      Point<dim> closest_point{};
+      this->closest_surface_point(evaluation_point, closest_point);
+
+      Point<dim> centered_point = this->align_and_center(evaluation_point);
+      if (superquadric(centered_point) > 0)
+        return (closest_point - evaluation_point).norm();
+      else
+        return -(closest_point - evaluation_point).norm();
+    }
+  else
+    return iterator->second;
+}
+
+template <int dim>
+double
+Superquadric<dim>::value_with_cell_guess(
+  const Point<dim> &evaluation_point,
+  const typename DoFHandler<dim>::active_cell_iterator /*cell*/,
+  const unsigned int /*component = 0*/)
+{
+  auto point_in_string = this->point_to_string(evaluation_point);
+  auto iterator        = this->value_cache.find(point_in_string);
+  if (iterator == this->value_cache.end())
+    {
+      // The closest point has to be found first, because it is used in value
+      // calculation.
+      Point<dim> closest_point{};
+      this->closest_surface_point(evaluation_point, closest_point);
+      Point<dim> copy_closest_point{};
+      for (unsigned int d = 0; d < dim; d++)
+        copy_closest_point[d] = closest_point[d];
+      this->closest_point_cache[point_in_string] = copy_closest_point;
+
+      double levelset                    = this->value(evaluation_point);
+      this->value_cache[point_in_string] = levelset;
+      return levelset;
+    }
+  else
+    return iterator->second;
+}
+
+template <int dim>
+std::shared_ptr<Shape<dim>>
+Superquadric<dim>::static_copy() const
+{
+  std::shared_ptr<Shape<dim>> copy = std::make_shared<Superquadric<dim>>(
+    this->half_lengths, this->exponents, this->position, this->orientation);
+  return copy;
+}
+
+template <int dim>
+void
+Superquadric<dim>::clear_cache()
+{
+  Shape<dim>::clear_cache();
+  this->closest_point_cache.clear();
```

Comment:
This should just be added directly to the original clear cache. There is no reason to have a different clear cache for different shapes.

Reply:
There is no closest point cache for other shapes. Should we add it you think?